### DISCUSSION TITLE
MissingGreenlet error even though the eager loading strategy by default

### DISCUSSION THREAD

**Discussion by MaximMukhametov** [sqlalchemy/sqlalchemy#9671] [Usage Questions] [answered]
Please tell me I can't figure out why the MissingGreenlet error occurs, even though the default is eager loading is "selectin" . And why if you close the session everything works fine?
```python
@dataclass
class ClientLimit:
    promo_id: UUID
    client_id: str
    value_spent_qty: int
    transaction_spent_qty: int
    id: int | None = None
    period_id: IntID | None = None
    period: ClientLimitPeriodProtocol | None = None
    
@dataclass
class Promo:
    name: str
    type: PromoType
    mechanic_type: MechanicsType
    id: UUID | None = None
```

<img width="1314" alt="Screenshot 2023-04-18 at 19 19 27" src="https://user-images.githubusercontent.com/59606832/232858046-1ee69369-11da-4d6a-ba99-07c58d62b86d.png">

**Comment by zzzeek** [comment, marked as answer]
this is a variation on the fourth bullet at https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession

> When constructing new objects, collections are always assigned a default, empty collection, such as a list in the above example:

>    A(bs=[], data="a2")


in this case I think it's Promo, this would solve the error, but doing that alone without doing other things wont give you the answer you want:

    Promo(client_limit = [])

if you don't do the above, then when you go to access promo.client_limit, it's unloaded, it has to load.

when you close the Session, promo is expunged, when you go to get() it again from the new Session, it loads it fresh and does the "lazy='selectin'"

The reason you can't do Promo(client_limit=[]) here without changing other things is because you are adding ClientLimit objects using foreign key identifiers, rather than just ClientLimit(promo=promo).   If you dont change that, then you **have** to see promo.client_limit reloaded to see the answer that you are looking for.

so to see that value you need to:

a. **do the get() with populate existing:**

    await session.get(Promo, id, populate_existing=True)

or 

b. **dont use FK identifiers to set up persistence, use the relationships, and assign a default collection**:

        p = Promo(client_limit = [])
        c1 = ClientLimit(promo=p)

or

        p = Promo(client_limit = [])
        p.client_limit.append(ClientLimit)
